Make GroupKFold use a stable sort

blitchj opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes: #29495


#### What does this implement/fix? Explain your changes.
Use stable sort in GroupKFold

#### Any other comments?
This does not reproduce the current results. But the current results aren't perfectly reproducible; however, they are likely to get the same sort on the same machine. It may be best to either consider this a breaking change as the new sort is different from the old. Or instead extend GroupKFold with a StableGroupKFold or add a stable parameter to GroupKFold.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

blitchj committed:
Make GroupKFold use a stable sort

adrinjalali commented:
Could you please add a test and a changelog entry?

blitchj commented:
> Could you please add a test and a changelog entry?

Sure I will do this once I give the better reproducible example in the linked issue that I've prepared.

Also before I do this I think it is worth discussing _if_ this should be done in this simple direct manor.  I think there is merit to allowing the user to set this, specifically because this fix does make the new sort repeatable, but if a user ran this before the fix on the same machine they probably got consistent results that will be different than what this will now output.

I would guess the best solution is 1 of 2 options:
1. Add a `stable` boolean option that defaults, at least for now, to False, with a possible warning, this way the default behavior is to continue to be consistent with the user's previous runs, but warn the users that it may not be as consistent as they expect if they run on another machine. And encourage new users to set it to True.
2. Add a new class StableGroupKFold search that only uses stable sort, but leave GroupKFold as it is, but add a warning at least in the docs.

Please let me know your thoughts as the maintainers of scikit-learn obviously have better opinions of how this type of issue should be handled within this package.

adrinjalali commented:
I would rather have this simply done as a FIX. But happy to hear what other @scikit-learn/core-devs think

adam2392 commented:
Yeah I don't see a reason we would ever want unstable sort in GroupKFold, but lmk if I'm missing anything.

Micky774 commented:
Ditto @adrinjalali, a FIX makes sense here imo.

blitchj committed:
unit test GroupKFold for stable search and update changelog

blitchj commented:
> @blitchj could you add a changelog entry as a `FIX` please?

Just updated this, hope it looks good.
Also added unit tests to enforce consistency in the Group assignment to be consistent with the new assignments based on stable search.

blitchj commented:
> ruff check --fix --output-format=full .

I ran this locally, seems non of the issues reference new code from this PR in the implementation or tests.

adam2392 commented:
> ruff check --fix --output-format=full .

You can run `ruff check ./sklearn/model_selection/ —fix` to fix the existing lint errors

blitchj committed:
linting, import order

adam2392 commented on doc/whats_new/v1.6.rst:
```diff
@@ -230,6 +230,10 @@ Changelog
   :pr:`29402` by :user:`Anurag Varma <Anurag-Varma>`.
 
+- |Fix| Make GroupKFold use a stable sort to ensure reproducability with ties in group size. 
+  This may result in different groupings than previous versions, but will now be reprodcable.
```
```suggestion
  This may result in different groupings than previous versions, but will now be reproducible across different machines and runs.
```

adam2392 commented on sklearn/model_selection/_split.py:
```diff
@@ -610,7 +610,7 @@ def _iter_test_indices(self, X, y, groups):
 
         # Distribute the most frequent groups first
-        indices = np.argsort(n_samples_per_group)[::-1]
+        indices = np.argsort(n_samples_per_group, kind="stable")[::-1]
```
Technically once we support numpy v2.0+ only, we could set this as `stable=True`.

https://numpy.org/doc/stable/reference/generated/numpy.argsort.html

Perhaps we can add an in-line comment:

```
# TODO: replace with np.argsort(n_samples_per_group, stable=True)[::-1] when numpy v2.0 is the minimum supported version
```

blitchj replied:
Sure, though, the `stable=True` docs on numpy 2.0 mention it just sets `kind="stable"` behind the scenes.

adam2392 replied:
Oh fair point. I missed that. Nvm then on this comment!

adam2392 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -2066,3 +2066,103 @@ def test_stratified_splitter_without_y(cv):
+        assert_array_equal(train, np.array(expected[i]))
+
+
+def test_group_k_fold_is_stable_no_tied():
```
```suggestion
def test_group_kfold_with_no_ties():
```

adam2392 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -2066,3 +2066,103 @@ def test_stratified_splitter_without_y(cv):
+        ),
+    ],
+)
+def test_group_k_fold_is_stable_tied(folds, n_groups, group_size, expected):
```
```suggestion
def test_group_kfold_is_stable_with_ties(folds, n_groups, group_size, expected):
```

adam2392 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -2066,3 +2066,103 @@ def test_stratified_splitter_without_y(cv):
         cv.split(X)
+
+
+@pytest.mark.parametrize(
+    "folds,n_groups,group_size,expected",
+    [
+        (
+            333,
+            333,
+            1,
+            [[j for j in range(333) if j != i] for i in reversed(range(333))],
+        ),
+        (
+            2,
+            41,
+            5,
+            [
+                list(
+                    chain.from_iterable(
+                        [range(10 * i + 5, 10 * i + 10) for i in range(20)]
+                    )
+                ),
+                list(
+                    chain.from_iterable([range(10 * i, 10 * i + 5) for i in range(21)])
+                ),
+            ],
+        ),
+    ],
+)
```
Is it possible to simplify this parametrization? Also, are all of them absolutely necessary?

blitchj replied:
Definitely can be reduced. The original intent for including both was:
1. This one is much easier to reason with what the desired output should be in a developers head. (it is long because for short sorts, I believe numpy is already locally stable)
2. This scenario actually uses a more reasonable setting for the sizes, and is based the on the minimum reproducible example from the issue. 

Only the second case is one I have shown to not be consistent previously.

blitchj replied:
If keeping only one, I prefer the second as I know it would not work previous to the fix. But the parameterization is over complicated with the whole itertool chain thing...

lesteve replied:
FWIW the second parametrized test does not fail on `main` on my machine.

I tried to come up with a simpler test and here is where I am at:
- I think this is simpler to check test set than train set. checking train against `[[j for j in range(333) if j != i] for i in reversed(range(333))]` is the same thing as checking test against `[i for i in reversed(range(333))]`
- at one point I thought OK one the 1 sample per group case I guess a `GroupKFold` and a `KFold` should give you the same thing except they don't, the test indices are reversed [0, 1, 2, ...] for `KFold` and `[332, 331, 330, ...]` for `GroupKFold`

adam2392 replied:
Both end up failing on my device, so either test is fine with me.

I think tho for maintainability of the test, the parameterization would be better if written like this:

```
def test_group_kfold_is_stable_with_ties(folds, n_groups, group_size):
    """Verify groups are assigned based on a stable sort, with ties."""
    # and should work if I change these arbitrarily 
    expected = function(folds, n_groups, group_size)
```

blitchj replied:
I'll give this a shot in a little bit, I like that formatting if I can make that expected function simple enough.
Also I was thinking a similar thing about checking the test set since it is simpliler logically.

blitchj replied:
Updated this test, I added an additional run with new parameters, to show the expected line is general. But can probably remove that as it is redundant for showing the GroupKFold is stable.

adam2392 reviewed (commented):
Thanks for the PR @blitchj! I left a few comments

blitchj committed:
Better unit test naming

Co-authored-by: Adam Li <adam2392@gmail.com>

blitchj committed:
additional better unit test naming

Co-authored-by: Adam Li <adam2392@gmail.com>

blitchj committed:
Changelog wording.

Co-authored-by: Adam Li <adam2392@gmail.com>

blitchj committed:
maketest_group_kfold_is_stable_with_ties unit test parameterization simpler

blitchj committed:
maketest_group_kfold_is_stable_with_ties unit test parameterization even simpler

adam2392 commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -2066,3 +2066,86 @@ def test_stratified_splitter_without_y(cv):
+
+    split = [a for a in gkf.split(X, groups=groups)]
+    for i, (_, test) in enumerate(split):
+        expected = X[groups[::-1] % folds == i]
```
Can you explain why `groups[::-1] % folds == i` is expected? Just to understand the logic here.

blitchj replied:
In `_iter_test_indices` the stable sort is reversed to get the largest first. When group lengths are all tied that is the same as just reversing the groups. Then the dealing of groups to folds will start with the first group (was last before reverse) and deal in order to each fol, and since argmin breaks ties with the earliest index it deals normally (not snake style)
Also this expected only works because the group labels are just their group index.

adam2392 replied:
I see! Can you explain this in the inline comment just for future reference?

lesteve replied:
It would be an equally good choice to have `expected = X[groups % folds == i]`. It would be more consistent with `KFold` in the `group_size=1` case https://github.com/scikit-learn/scikit-learn/pull/29496#discussion_r1684159533 i.e. in the first split the test set start contains the first sample of X. 

Do we want to do this since we are changing the splits of `GroupKFold`, I would say maybe?

blitchj replied:
I think that is definitely a valid expectation and this wojld be a good time to change that if it will be changed. That way the assignments can maintain consistency across versions, barring required changes. It would also require an update to the no ties test I wrote.

If you all think that is a good change, I can update the assignment loop to be reversed to solve this and make a comment as to why.

blitchj committed:
comment test_group_kfold_is_stable_with_ties expected

lesteve commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -2066,3 +2066,87 @@ def test_stratified_splitter_without_y(cv):
+        assert_array_equal(test, expected)
+
+
+def test_group_kfold_with_no_ties():
```
This test does not fail on `main` which I guess is expected if there are no ties, it does not matter whether the sort is stable or not.

Can you explain a bit what this additional test brings, for example maybe it covers a blind spot of the existing tests?

blitchj replied:
I believe it covers a blindspot, about consistency of the assignment order. It is designed to catch any implementation changes to GroupKFold that would break the consistency of the splits between versions.

However, if sklearn doesn't want to guarantee consistency of the group assignments between version s then this test would be unnecessary.

adrinjalali commented:
also, xref re performance: https://github.com/scikit-learn/scikit-learn/issues/29531#issuecomment-2242827339

blitchj commented:
#29531 Has moved to `Needs Discussion`. I wonder if this should be part of that discussion or not? 
I would think GroupKFold will ~usually~ have more groups to sort than RFE has feature importances to sort, but I would also think there are more often many ties in GroupKFold, either due to use cases where all / most groups have a consistent size, or just due to group sizes being integers while feature importances are floats. Additionally, I would think this one has less concerns about randomizing ties and more reason to control for consistency.
